File: src/main/java/com/fewlaps/quitnowcache/QNCache.java
package com.fewlaps.quitnowcache;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class QNCache<T> {

    public static final long KEEPALIVE_FOREVER = 0;

    private final ConcurrentHashMap<String, QNCacheBean<T>> cache;
    private boolean caseSensitiveKeys = true;
    private Integer autoReleaseInSeconds;
    private Long defaultKeepaliveInMillis;
    private DateProvider dateProvider = DateProvider.SYSTEM;

    public QNCache(boolean caseSensitiveKeys, Integer autoReleaseInSeconds, Long defaultKeepaliveInMillis) {
        this.caseSensitiveKeys = caseSensitiveKeys;
        if (autoReleaseInSeconds != null && autoReleaseInSeconds > 0) {
            this.autoReleaseInSeconds = autoReleaseInSeconds;
        }
        if (defaultKeepaliveInMillis != null && defaultKeepaliveInMillis > 0) {
            this.defaultKeepaliveInMillis = defaultKeepaliveInMillis;
        }

        cache = new ConcurrentHashMap<String, QNCacheBean<T>>();

        startAutoReleaseServiceIfNeeded();
    }

    private void startAutoReleaseServiceIfNeeded() {
        if (autoReleaseInSeconds != null) {
            ScheduledExecutorService ses = Executors.newSingleThreadScheduledExecutor();

            ses.scheduleAtFixedRate(new Runnable() {
                @Override
                public void run() {
                    purge();
                }
            }, autoReleaseInSeconds, autoReleaseInSeconds, TimeUnit.SECONDS);
        }
    }

    boolean isCaseSensitiveKeys() {
        return caseSensitiveKeys;
    }

    Integer getAutoReleaseInSeconds() {
        return autoReleaseInSeconds;
    }

    Long getDefaultKeepaliveInMillis() {
        return defaultKeepaliveInMillis;
    }

    private long now() {
        return dateProvider.now();
    }

    protected void setDateProvider(DateProvider dateProvider) {
        this.dateProvider = dateProvider;
    }

    public void set(String key, T value) {
        if (defaultKeepaliveInMillis != null) {
            set(key, value, defaultKeepaliveInMillis);
        } else {
            set(key, value, KEEPALIVE_FOREVER);
        }
    }

    public void set(String key, T value, long keepAliveInMillis) {
        String effectiveKey = getEffectiveKey(key);

        if (keepAliveInMillis >= 0) {
            cache.put(effectiveKey, new QNCacheBean<T>(value, now(), keepAliveInMillis));
        }
    }

    public void set(String key, T value, long keepAliveUnits, TimeUnit timeUnit) {
        set(key, value, timeUnit.toMillis(keepAliveUnits));
    }

    /**
     * Gets an element from the cache.
     */
    public T get(String key) {
        String effectiveKey = getEffectiveKey(key);

        QNCacheBean<T> retrievedValue = cache.get(effectiveKey);
        if (retrievedValue == null || !retrievedValue.isAlive(now())) {
            return null;
        } else {
            return retrievedValue.getValue();
        }
    }

    /**
     * Gets an element from the cache. If the element exists but it's dead,
     * it will be removed from the cache to free memory. It could call
     * an internal synchronized method, so avoid calling this method if
     * it's not needed.
     */
    public T getAndPurgeIfDead(String key) {
        String effectiveKey = getEffectiveKey(key);

        QNCacheBean<T> retrievedValue = cache.get(effectiveKey);
        if (retrievedValue == null) {
            return null;
        } else if (retrievedValue.isAlive(now())) {
            return retrievedValue.getValue();
        } else {
            cache.remove(effectiveKey);
            return null;
        }
    }

    public void remove(String key) {
        String effectiveKey = getEffectiveKey(key);
        cache.remove(effectiveKey);
    }

    /**
     * Removes all the elements of the cache, ignoring if they're dead or alive
     */
    public void clear() {
        cache.clear();
    }

    public List<String> keySetDeadAndAlive() {
        return Collections.list(cache.keys());
    }

    public List<String> keySetAlive() {
        List<String> aliveKeys = new ArrayList<String>();

        for (String key : keySetDeadAndAlive()) {
            if (isKeyAlive(key)) {
                aliveKeys.add(key);
            }
        }

        return aliveKeys;
    }

    public List<String> keySetDead() {
        List<String> aliveKeys = new ArrayList<String>();

        for (String key : keySetDeadAndAlive()) {
            if (isKeyDead(key)) {
                aliveKeys.add(key);
            }
        }

        return aliveKeys;
    }

    public List<String> keySetStartingWith(String keyStartingWith) {
        List<String> filteredKeys = new ArrayList<String>();
        String effectiveKeyStartingWith = getEffectiveKey(keyStartingWith);

        for (String key : keySetDeadAndAlive()) {
            if (key.startsWith(effectiveKeyStartingWith)) {
                filteredKeys.add(key);
            }
        }

        return filteredKeys;
    }

    public List<String> keySetAliveStartingWith(String keyStartingWith) {
        List<String> aliveFilteredKeys = new ArrayList<String>();

        for (String key : keySetStartingWith(keyStartingWith)) {
            if (isKeyAlive(key)) {
                aliveFilteredKeys.add(key);
            }
        }

        return aliveFilteredKeys;
    }

    public boolean isKeyAlive(String key) {
        String effectiveKey = key;
        return cache.get(effectiveKey).isAlive(now());
    }

    public boolean isKeyDead(String key) {
        return !isKeyAlive(key);
    }

    /**
     * Counts how much alive elements are living in the cache
     */
    public int size() {
        return sizeAliveElements();
    }

    /**
     * Counts how much alive elements are living in the cache
     */
    public int sizeAliveElements() {
        int size = 0;

        for (QNCacheBean<T> cacheValue : cache.values()) {
            if (cacheValue.isAlive(now())) {
                size++;
            }
        }
        return size;
    }

    /**
     * Counts how much dead elements exist in the cache
     */
    public int sizeDeadElements() {
        return cache.size() - sizeAliveElements();
    }

    /**
     * Counts how much elements are living in the cache, ignoring if they are dead or alive
     */
    public int sizeDeadAndAliveElements() {
        return cache.size();
    }

    /**
     * The common isEmpty() method, but only looking for alive elements
     */
    public boolean isEmpty() {
        return sizeAliveElements() == 0;
    }

    /**
     * The common contains() method
     */
    public boolean contains(String key) {
        String effectiveKey = getEffectiveKey(key);
        return get(effectiveKey) != null;
    }

    /**
     * Removes the dead elements of the cache, to free memory
     */
    void purge() {
        Iterator<Map.Entry<String, QNCacheBean<T>>> it = cache.entrySet().iterator();

        while (it.hasNext()) {
            Map.Entry<String, QNCacheBean<T>> entry = it.next();
            QNCacheBean<T> bean = entry.getValue();
            if (!bean.isAlive(now())) {
                it.remove();
            }
        }
    }

    /**
     * If caseSensitiveKeys is false, it returns a key in lowercase. It will be
     * the key of all stored values, so the cache will be totally caseinsensitive
     */
    private String getEffectiveKey(String key) {
        if (!caseSensitiveKeys) {
            return key.toLowerCase();
        }
        return key;
    }
}

File: src/main/java/com/fewlaps/quitnowcache/QNCache.kt
package com.fewlaps.quitnowcache

import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit

class QNCache<T>(
        val caseSensitiveKeys: Boolean = true,
        val autoReleaseInSeconds: Int = WITHOUT_AUTORELEASE,
        val defaultKeepaliveInMillis: Long = KEEPALIVE_FOREVER) {

    private val cache: ConcurrentHashMap<String, QNCacheBean<T>> = ConcurrentHashMap()

    private var executorService: ScheduledExecutorService? = null

    var dateProvider = DateProvider.SYSTEM!!

    init {
        startAutoReleaseServiceIfNeeded()
    }

    /**
     * The common isEmpty() method, but only looking for alive elements
     */
    fun isEmpty() = sizeAliveElements() == 0

    fun shutdown() {
        clear()
        executorService?.shutdown()
    }

    private fun startAutoReleaseServiceIfNeeded() {
        if (autoReleaseInSeconds > 0) {
            executorService = Executors.newSingleThreadScheduledExecutor()
            executorService?.scheduleAtFixedRate({ purge() }, autoReleaseInSeconds.toLong(), autoReleaseInSeconds.toLong(), TimeUnit.SECONDS)
        }
    }

    operator fun set(key: String, value: T) {
        set(key, value, defaultKeepaliveInMillis)
    }

    operator fun set(key: String, value: T, keepAliveUnits: Long, timeUnit: TimeUnit) {
        set(key, value, timeUnit.toMillis(keepAliveUnits))
    }

    operator fun set(key: String, value: T, keepAliveInMillis: Long) {
        if (keepAliveInMillis >= 0) {
            cache[getEffectiveKey(key)] = QNCacheBean(value, now(), keepAliveInMillis)
        }
    }

    /**
     * Gets an element from the cache. If it's null, the default value will be returned instead.
     */
    fun getOrDefault(key: String, defaultValue: T): T = get(key) ?: defaultValue

    /**
     * Gets an element from the cache.
     */
    operator fun get(key: String) = get(key, purgeIfDead = false)

    /**
     * Gets an element from the cache. If the element exists but is dead,
     * it will be removed from the cache to free memory. It could call
     * an internal synchronized method, so avoid calling this method if
     * you are not storing huge objects in terms of memory.
     */
    fun getAndPurgeIfDead(key: String) = get(key, purgeIfDead = true)

    private fun get(key: String, purgeIfDead: Boolean): T? {
        val effectiveKey = getEffectiveKey(key)
        val retrievedValue = cache[effectiveKey]

        return when {
            retrievedValue == null -> null
            retrievedValue.isAlive(now()) -> retrievedValue.value
            else -> {
                if (purgeIfDead) {
                    cache.remove(effectiveKey)
                }
                null
            }
        }
    }

    fun remove(key: String) {
        val effectiveKey = getEffectiveKey(key)
        cache.remove(effectiveKey)
    }

    /**
     * Removes all the elements of the cache, ignoring if they're dead or alive
     */
    fun clear() = cache.clear()

    fun keySet() = keySetAlive()

    fun keySetDeadAndAlive(): List<String> = cache.keys().toList()

    fun keySetAlive(): List<String> {
        return keySetDeadAndAlive().filter {
            isKeyAlive(it)
        }
    }

    fun keySetDead(): List<String> {
        return keySetDeadAndAlive().filter {
            isKeyDead(it)
        }
    }

    fun keySetStartingWith(start: String?): List<String> {
        if (start == null) return Collections.emptyList()

        val effectiveKeyStartingWith = getEffectiveKey(start)

        return keySetDeadAndAlive().filter {
            it.startsWith(effectiveKeyStartingWith)
        }
    }

    fun keySetAliveStartingWith(start: String?): List<String> {
        if (start == null) return Collections.emptyList()

        return keySetStartingWith(start).filter {
            isKeyAlive(it)
        }
    }

    fun isKeyAlive(key: String): Boolean {
        val value = cache[key] ?: return false
        return value.isAlive(now())
    }

    fun isKeyDead(key: String) = !isKeyAlive(key)

    /**
     * Counts how much alive elements are living in the cache
     */
    fun size() = sizeAliveElements()

    /**
     * Counts how much alive elements are living in the cache
     */
    fun sizeAliveElements(): Int {
        return cache.values.filter { it.isAlive(now()) }.count()
    }

    /**
     * Counts how much dead elements exist in the cache
     */
    fun sizeDeadElements() = cache.size - sizeAliveElements()

    /**
     * Counts how much elements are living in the cache, ignoring if they are dead or alive
     */
    fun sizeDeadAndAliveElements() = cache.size

    /**
     * The common contains() method that looks for alive elements
     */
    operator fun contains(key: String) = get(getEffectiveKey(key)) != null

    /**
     * Removes the dead elements of the cache to free memory
     */
    fun purge() = cache.entries.removeIf { isKeyDead(it.key) }

    /**
     * If caseSensitiveKeys is false, it returns a key in lowercase. It will be
     * the key of all stored values, so the cache will be totally caseinsensitive
     */
    private fun getEffectiveKey(key: String): String {
        return when {
            caseSensitiveKeys -> key
            else -> key.toLowerCase()
        }
    }

    private fun now(): Long {
        return dateProvider.now()
    }

    class Builder {
        private var caseSensitiveKeys = true
        private var autoReleaseInSeconds: Int = QNCache.WITHOUT_AUTORELEASE
        private var defaultKeepaliveInMillis = QNCache.KEEPALIVE_FOREVER

        fun caseSensitiveKeys(caseSensitiveKeys: Boolean): Builder {
            this.caseSensitiveKeys = caseSensitiveKeys
            return this
        }

        fun autoRelease(units: Int, timeUnit: TimeUnit): Builder {
            this.autoReleaseInSeconds = java.lang.Long.valueOf(timeUnit.toSeconds(units.toLong())).toInt()
            return this
        }

        fun autoReleaseInSeconds(autoReleaseInSeconds: Int): Builder {
            this.autoReleaseInSeconds = autoReleaseInSeconds
            return this
        }

        fun defaultKeepalive(units: Int, timeUnit: TimeUnit): Builder {
            this.defaultKeepaliveInMillis = timeUnit.toMillis(units.toLong())
            return this
        }

        fun defaultKeepaliveInMillis(defaultKeepaliveInMillis: Long): Builder {
            this.defaultKeepaliveInMillis = defaultKeepaliveInMillis
            return this
        }

        fun <T> build(): QNCache<T> {
            if (autoReleaseInSeconds < 0) {
                this.autoReleaseInSeconds = WITHOUT_AUTORELEASE
            }
            if (defaultKeepaliveInMillis < 0) {
                this.defaultKeepaliveInMillis = KEEPALIVE_FOREVER
            }
            return QNCache(caseSensitiveKeys, autoReleaseInSeconds, defaultKeepaliveInMillis)
        }
    }

    companion object {
        const val WITHOUT_AUTORELEASE: Int = 0
        const val KEEPALIVE_FOREVER: Long = 0
    }
}